ReactSkia/core_modules/RSkSpatialNavigator.cpp

```diff
@@ -0,0 +1,380 @@
+    RNS_LOG_DEBUG("Send : " << eventType  << " To ComponentTag : " << tag );
+    NotificationCenter::defaultCenter().emit("RCTTVNavigationEventNotification",
+                                                folly::dynamic(folly::dynamic::object("eventType", eventType.c_str())
+                                                                              ("eventKeyAction", (int)keyAction)
+                                                                              ("tag", tag)
+                                                                              ("target", tag)
+                                                                              ));
+}
+
+/*
+ * Add a new candidate to the Navigatable if it meets the following criteria:
+ * isTVSelectable, accessible are true
+*/
+void RSkSpatialNavigator::addToNavList(std::shared_ptr<RSkComponent> newCandidate) {
+    Component newCandData = newCandidate.get()->getComponentData();
+#if 0
+    if (candidate.props->accessible == true || candidate.isTVSelectable == true ) {
+#else
+    if(1) {
+#endif
+        navComponentList_.push_back(newCandidate.get());
+        // TODO If this new component has hasTVPreferredFocus then send Blur to current focused element and send focus to this new candidate
+        if(1) { //newCandidate.get()->props.hasTVPreferredFocus == true) {
+            if(currentFocus_)
+                sendNotificationWithEventType("blur", currentFocus_->getComponentData().tag);
+
+            sendNotificationWithEventType("Focus", newCandData.tag);
+            currentFocus_ = newCandidate.get();
+        }
+    }
+}
+
+void RSkSpatialNavigator::removeFromNavList(std::shared_ptr<RSkComponent> candidate) {
+    Component canData = candidate.get()->getComponentData();
+#if 0
+    if (candidate.props->accessible == true || candidate.isTVSelectable == true ) {
+#else
+    if(1) {
+#endif
+        CandidateList::iterator it;
+        it = std::find(navComponentList_.begin(), navComponentList_.end(), candidate.get());
+        if (it != navComponentList_.end()) {
+            // Reset data if candidate being removed is currently focused
+            if (currentFocus_ == candidate.get()) {
+                currentFocus_ = nullptr;
+            }
+            navComponentList_.erase(it);
+        }
+    }
+}
+
+void RSkSpatialNavigator::updateInNavList(std::shared_ptr<RSkComponent> candidate) {
+    Component canData = candidate.get()->getComponentData();
+
+    CandidateList::iterator it;
+    it = std::find(navComponentList_.begin(), navComponentList_.end(), candidate.get());
+
+    if (it != navComponentList_.end()) {
+        // Candidate found in the navigatable list, Reset data if candidate's focusable props have changed
+        // TODO: Check the interdependency / priority of the props below
+        if (0/*canData.accessible == false || canData.isTVSelectable == false*/) {
+            if (currentFocus_ == candidate.get()) {
+                currentFocus_ = nullptr;
```

Comment:
since we are resetting the currentFocus_ ; should there be a blur event sent.
the currentFocus_ element would be in "focus" and in addNav* currentFocus_ is checked and if new element add "focus" event is sent. one "blur" event is missed.

Reply:
Done.